### PR TITLE
Add codechecks reporting

### PR DESCRIPTION
**Pull request by cgewecke**
[EDITED]
@krzkaczor wrote and told me about this - it's a [CI service](https://codechecks.io/) that makes it possible to track gas usage changes across pull requests, like coveralls.  Amazing. 

### User setup
+ Enable project on [codechecks.io](https://codechecks.io/)
+ Install codechecks as a dev dependency: `npm install --save-dev @codechecks/client`
+  Add a `codechecks.yml` to their project as below:
```yml
checks:
  - name: eth-gas-reporter/codechecks
```
+ Run `codechecks` as a step in their CircleCI build
```yml
steps:
  - checkout
  - run: yarn install
  - run: yarn test
  - run: yarn codechecks
```

### CI Screenshot
Screen Shot 2019-06-18 at 12 25 49 PM

### Graph output
```diff
......................|..................................|.............|............................·
    Solc: v0.5.0      ·        Optimized: false          ·  Runs: 200  ·    Block: 8000000 gas      
······················|··································|·············|·····························
  METHODS             ·                  1 gwei/gas                    ·      237.80 eur/eth        
················|·····|·······················|··········|·············|··············|··············
  Contract      ·     ·  Method               ·   Gas    ·    Diff     ·  # calls     ·  eur (avg)  
················|·····|·······················|··········|·············|··············|··············
  EtherRouter         ·  setResolver          ·   43192  ·          0  ·           1  ·       0.01  
······················|·······················|··········|·············|··············|··············
  Factory             ·  deployVersionB       ·  107123  ·          0  ·           1  ·       0.03  
······················|·······················|··········|·············|··············|··············
  MetaCoin            ·  sendCoin             ·   51019  ·          0  ·           1  ·       0.01  
······················|·······················|··········|·············|··············|··············
  Migrations          ·  setCompleted         ·   27034  ·          0  ·           6  ·       0.01  
······················|·······················|··········|·············|··············|··············
  MultiContractFileA  ·  hello                ·   41419  ·          0  ·           1  ·       0.01  
······················|·······················|··········|·············|··············|··············
  MultiContractFileB  ·  goodbye              ·   41419  ·          0  ·           1  ·       0.01  
······················|·······················|··········|·············|··············|··············
- Resolver            ·  register             ·   37633  ·         32  ·           2  ·       0.01  
······················|·······················|··········|·············|··············|··············
  VariableCosts       ·  addToMap             ·   90341  ·          0  ·           7  ·       0.02  
······················|·······················|··········|·············|··············|··············
  VariableCosts       ·  methodThatThrows     ·   41599  ·          0  ·           2  ·       0.01  
······················|·······················|··········|·············|··············|··············
  VariableCosts       ·  otherContractMethod  ·   57407  ·          0  ·           2  ·       0.01  
······················|·······················|··········|·············|··············|··············
  VariableCosts       ·  removeFromMap        ·   36481  ·          0  ·           8  ·       0.01  
······················|·······················|··········|·············|··············|··············
  VariableCosts       ·  sendPayment          ·   32335  ·          0  ·           1  ·       0.01  
······················|·······················|··········|·············|··············|··············
- VariableCosts       ·  setString            ·   86198  ·        768  ·           2  ·       0.02  
······················|·······················|··········|·············|··············|··············
  VariableCosts       ·  transferPayment      ·   32186  ·          0  ·           1  ·       0.01  
······················|·······················|··········|·············|··············|··············
  VersionA            ·  setValue             ·   25674  ·          0  ·           2  ·       0.01  
······················|·······················|··········|·············|··············|··············
  Wallet              ·  sendPayment          ·   32181  ·          0  ·           1  ·       0.01  
······················|·······················|··········|·············|··············|··············
  Wallet              ·  transferPayment      ·   32164  ·          0  ·           1  ·       0.01  
······················|·······················|··········|·············|··············|··············
  DEPLOYMENTS         ·                                                ·  % of limit  ·             
······················|·······················|··········|·············|··············|··············
  ConvertLib                                  ·  111791  ·          0  ·       1.7 %  ·       0.03  
··············································|··········|·············|··············|··············
  EtherRouter                                 ·  278020  ·          0  ·       4.1 %  ·       0.07  
··············································|··········|·············|··············|··············
  Factory                                     ·  324331  ·          0  ·       4.8 %  ·       0.08  
··············································|··········|·············|··············|··············
+ MetaCoin                                    ·  358508  ·        -64  ·       5.3 %  ·       0.09  
··············································|··········|·············|··············|··············
  Migrations                                  ·  284908  ·          0  ·       4.2 %  ·       0.07  
··············································|··········|·············|··············|··············
  MultiContractFileA                          ·   90745  ·          0  ·       1.4 %  ·       0.02  
··············································|··········|·············|··············|··············
  MultiContractFileB                          ·   90745  ·          0  ·       1.4 %  ·       0.02  
··············································|··········|·············|··············|··············
  Resolver                                    ·  430644  ·          0  ·       6.4 %  ·       0.10  
··············································|··········|·············|··············|··············
- VariableConstructor                         ·  987116  ·      14096  ·      14.7 %  ·       0.23  
··············································|··········|·············|··············|··············
  VariableCosts                               ·  930528  ·          0  ·      13.8 %  ·       0.22  
··············································|··········|·············|··············|··············
  VersionA                                    ·   88665  ·          0  ·       1.3 %  ·       0.02  
··············································|··········|·············|··············|··············
  Wallet                                      ·  217795  ·          0  ·       3.2 %  ·       0.05  
..............................................|..........|.............|..............|.............·
```
Some resources:
+ [truffle-codechecks](https://github.com/krzkaczor/truffle-codechecks)
+ Integration testing for this PR at the [gas-check repo](https://github.com/cgewecke/gas-check)
+ codechecks [docs](https://github.com/codechecks/docs)

### CODE REVIEW

**Comment by krzkaczor** [comment]
Great work @cgewecke! I see that you made great progress already! 🎉 I would appreciate any feedback regarding codechecks’ docs and client part.  One thing that I noticed is that you decided to integrate with codechecks (CC) client directly.

The canonical way of using CC is to create a separate codecheck (separate package). This works great in many cases but I get why you decided on a little bit different approach in this case. Since recording “the artifact” (JSON describing gas execution costs) happens inside truffle one might integrate CC directly inside this process. This makes sense and results in a very pleasant setup for the end user but the programmatic access like this is not yet nicely supported. That’s why you needed to create `codechecks.js` on the fly etc — sorry about that. I can imagine that executing something like `const cc = codechecks.init()` on the CC client and having ready to use instance would make things easier for you.

I am gonna prioritize improving API for this use case as well as describing it in the docs. If you have any comments or ideas regarding this just let me know :)

**Comment by cgewecke** [comment]
@krzkaczor Hi, thanks for looking this over! 

> The canonical way of using CC is to create a separate codecheck (separate package). 

Yes, this makes sense - I've made some revisions to the reporter so it just writes a JSON data to the root folder whenever `process.env.CI` is true. A separate codecheck should be able to consume it whenever it's present. FWIW I think your design is good.

I *am* having two problems in my [test repo](https://github.com/cgewecke/gas-check) however.
+ The initial PR test run errors on the `isPR` check. You can see an example of this [here](https://circleci.com/gh/cgewecke/gas-check/46?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link). It looks like `CIRCLE_PULL_REQUEST` for that event is `undefined` but subsequent commits are fine. (super weird)
+ I cannot load my historical data using `getValue`. If you look through the [gas-check](https://github.com/cgewecke/gas-check/pulls) CI builds on the `yarn codechecks` step, you can see it looks like data is being saved correctly but retrieval always yields null.

If you have a chance could you peek at the backend logs and see what I'm doing wrong here? Sorry, thanks. 
 
> I would appreciate any feedback regarding codechecks’ docs and client part.

Cool, yes I will open some issues over there (at the landing repo?)

**Comment by krzkaczor** [comment]
> The initial PR test run errors on the isPR check. You can see an example of this here. It looks like CIRCLE_PULL_REQUEST for that event is undefined but subsequent commits are fine. (super weird)

I think I know exactly why it happens. When you pushed your initial commit, the PR was still not created (you just clicked on the UI right after pushing). This is a well known problem to me and I am playing with some fixes (like a setting to always speculate that its gonna be merged to master (if its not executed on master)).

> I cannot load my historical data using getValue. If you look through the gas-check CI builds on the yarn codechecks step, you can see it looks like data is being saved correctly but retrieval always yields null.

Are u aware that getting a value is always happening from the base branch? so probably master? are u sure that there was a baseline recorded? I saw some commits chaning namespaces so maybe it used the old key?

> Apologies I think I finally understand this :) - codecheck.js should save newly collected data on the merge build and that's how you get the correct base commit hash?

YES. Your codecheck should *always* record data then when its executed on master it will just record the baseline.

**Comment by krzkaczor** [comment]
And another note: if you like the "canonical" design then I would suggest that this package shouldn't know anything about codechecks at all. The only change here would be the ability to output report as a JSON. Then storing, retrieving and diffing those reports should be part of the (let's say) `eth-gas-reporter-codechecks` package.

**Comment by cgewecke** [comment]
@krzkaczor Do you mostly use isPR to get the base commit? One thing I was thinking is that it might be possible to get this info from the final [PR webhook event](https://developer.github.com/v3/activity/events/types/#pullrequestevent) which contains some merge related stuff. This might complicate things on the back end for you though....

> Then storing, retrieving and diffing those reports should be part of the (let's say) eth-gas-reporter-codechecks package.

Yes! Another possibility is that the codechecks.js logic ships with eth-gas-reporter, but it's accessed via the `codecheck.yml`. 
```yml
checks:
- name: eth-gas-reporter/codechecks
```

**Comment by krzkaczor** [comment]
> Do you mostly use isPR to get the base commit? One thing I was thinking is that it might be possible to get this info from the final PR webhook event which contains some merge related stuff. This might complicate things on the back end for you though....

I am not sure if I understand correctly but since CC runs as part of CI (which is usually triggered by commits) you can't really use webhooks to get additional info (or its just simpler to ask API directly). What you can do is just simply ask Github API (we proxy it through our backend) what's the status of this commit we already do something along these lines. 

The problem is not merging back to master since CI is always triggered on master branch. The problem is the first commit on a new branch when PR is not yet created.


> Yes! Another possibility is that the codechecks.js logic ships with eth-gas-reporter, but it's accessed via the codecheck.yml.

`eth-gas-reporter/codechecks` is pretty neat actually 👍

**Comment by krzkaczor** [comment]
Btw. i think you can use emojis only in `longDescription` :(

**Comment by cgewecke** [comment]
@krzkaczor If on the first commit the env variables look like this:
```shell
CIRCLE_PULL_REQUEST: undefined
CIRCLE_SHA1: 457484e81a2e942b305ecdbd5f5e4d30ba0ae836
CIRCLE_PROJECT_USERNAME: cgewecke
CIRCLE_PROJECT_REPONAME: gas-check
```
is it possible to run a [GH request](https://developer.github.com/v3/pulls/) through the back end for a list of open PRs to see if that commit hash is present and get the PR id that way?  So strange that field isn't filled in...how is it possible that the PR isn't created yet? I would guess that Circle starts running **because** it's been triggered by a PullRequestEvent the repo is subscribed to on their side.   

> The problem is not merging back to master since CI is always triggered on master branch.

Ok, yes this is reasonable. I have a small worry that sometimes (esp on Travis) people turn off the push builds because they're useless and result in longer wait queues. If the codechecks back end subscribed to PR events for a repo perhaps it could get the merge commit hash even in these cases.

**Comment by cgewecke** [comment]
@krzkaczor I'm pretty happy with this, will probably merge later today and ping you over at codechecks with misc issues :) Thanks so much for all your help and if you have any other suggestions here please LMK.

File: package.json [hunk 1]
```diff
@@ -31,6 +31,7 @@
   },
   "homepage": "https://github.com/cgewecke/eth-gas-reporter#readme",
   "dependencies": {
+    "@codechecks/client": "^0.1.0",
```
Comment (krzkaczor) [inline package.json]: this should be a peer dependency I think.

Comment (krzkaczor) [reply package.json]: and possibly dev dep

**Comment by krzkaczor** [comment]
I think you ended up with really nice implementation!  I left a few notes.

In future, I was thinking that probably there should be a shared directory (like `.codechecks-workspace` that should be always get ignored and can be used as a standarized way of passing such artifacts (reports) locally as well as for some other things. This would be backed into a client itself with helpers etc. But that's a low priority. 

> is it possible to run a GH request through the back end for a list of open PRs to see if that commit hash is present and get the PR id that way? 

Yes, it is but imagine a situation like this. First of all, few assumptions:
 - CI is always triggered only on commits (at least in circleci)
 - CC is always executed on CI 

Now, imagine that you commit to your new branch and forget about your code. You're back in a week and create PR. CircleCI won't run CI on that commit again. Even if you would create PR 5 ms later it would still detect as running as part of a PR but rather a separate commit.

That's why I want to implement this "speculative" base branch feature that would basically always try to guess PR branch and act like it was a PR.

> I'm pretty happy with this, will probably merge later today and ping you over at codechecks with misc issues

Great! We can chat on email again. Let me know when you're ready with merging this so I can promote your work on my social channels as well :)

File: package.json (hunk 1 above)
Comment (cgewecke) [reply package.json]: Ah yes! Thanks.

**Comment by cgewecke** [comment]
@krzkaczor Ah I see what you mean about commits in CI now . . . it's tricky. 

> .codechecks-workspace

👍 

I *think* this will publish next Monday morning (GMT -7). This feature needs some docs work and there are a couple other things I'd like to put in that release.

Will ping [JoinColony](https://github.com/JoinColony/colonyNetwork) when this is out to see if they are interested in trying it b/c they currently use the gas reporter in CIrcleCI.

**Comment by krzkaczor** [comment]
That's amazing! I will try today to migrate `truffle-codechecks` to `eth-gas-reporter` codecheck.

**Comment by krzkaczor** [comment]
I just migrated my example repo https://github.com/krzkaczor/truffle-codechecks/pull/2 

I need to admit that it works very nice! In my previous solution user would need to track gas usage explicitly which sucked, now it's automatic for all methods — amazing.

Few suggestions:
* are u aware that you can run codechecks locally? In this case, it will most likely throw an error that `gasReporterOutput.json` doesn't exist. Probably there should be more detailed information that `gasReporterOutput.json` couldn't be found and you can create it by running `CI=true npm test`
* did you noticed that `longDescription` supports Github markdown? Currently generated table uses a different format and that's why it's not rendered nicely.

**Comment by cgewecke** [comment]
@krzkaczor 

> it will most likely throw an error that gasReporterOutput.json doesn't exist

Oh thanks, I'll fix this.

> did you noticed that longDescription supports Github markdown

Yes. Agree that the terminal table report is (much) uglier than rendered markdown table. The only benefit is that it can use markdown diff syntax highlighting to generate the red/green lines which pick out the changes. There seems to be no way of adding color to text in a markdown table (except emoji). Probably gas reporter should support both formats - there are pros and cons to each.

It might be cool if there was a helper which let you pass data to a library like d3.js, have it rendered by headless chromium, converted to svg and hosted somewhere as an image that Github will automatically display in the long description part of checks tab. This would make it possible to turn arbitrary info into colored displays. 

Believe chromium ships in a docker image and is small enough to run inside the Circle container, not sure There might be a simpler way of accomplishing the same thing.

**Comment by krzkaczor** [comment]
Ahh, I get you. Yeah diff support is nice to have. Other option would be to use emojis to make changes easily visible. Anyway, I think that the current solution is great for start.

**Pull request merged**